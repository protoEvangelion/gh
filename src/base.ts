/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as configs from './configs'
import { safeReaddir } from './fp'
import * as Future from 'fluture'

// -- Config -------------------------------------------------------------------

export function clone(o) {
    return JSON.parse(JSON.stringify(o))
}

// -- Utils --------------------------------------------------------------------

/**
 * Returns files in a folder path that match a given patter
 */
export function find(
    dirPath: string,
    optPattern = /.*/
): Future.FutureInstance<NodeJS.ErrnoException, string[]> {
    return safeReaddir(dirPath).map(dirs => {
        return dirs.filter(file => {
            return optPattern.test(file)
        })
    })
}

export function getUser() {
    return configs.getConfig().github_user || process.env.GH_USER
}

// Export some config methods to allow plugins to access them
export const getConfig = configs.getConfig
export const writeGlobalConfig = configs.writeGlobalConfig
