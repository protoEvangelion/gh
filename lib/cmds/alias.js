/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: CONTRIBUTORS)
 * SPDX-License-Identifier: BSD-3-Clause
 */

'use strict'

// -- Requires -------------------------------------------------------------------------------------
var base = require('../base'),
    configs = require('../configs'),
    logger = require('../logger'),
    config = base.getConfig()

// -- Constructor ----------------------------------------------------------------------------------

function Alias(options) {
    this.options = options
}

// -- Constants ------------------------------------------------------------------------------------

Alias.DETAILS = {
    alias: 'al',
    description: 'Create alias for a username.',
    commands: ['add', 'list', 'remove'],
    options: {
        add: String,
        list: Boolean,
        remove: String,
        user: String,
    },
    shorthands: {
        a: ['--add'],
        l: ['--list'],
        r: ['--remove'],
        u: ['--user'],
    },
    payload: function(payload, options) {
        if (payload[0]) {
            options.add = payload[0]
            options.user = payload[1]
        } else {
            options.list = true
        }
    },
}

// -- Commands -------------------------------------------------------------------------------------

Alias.prototype.run = function() {
    var instance = this,
        options = instance.options

    if (options.add) {
        if (!options.user) {
            logger.error('You must specify an user, try --user username.')
        }

        logger.debug('Creating alias ' + options.add)
        instance.add()
    }

    if (options.list) {
        instance.list(function(err, data) {
            var item

            for (item in data) {
                if (data.hasOwnProperty(item)) {
                    logger.log(logger.colors.cyan(item) + ': ' + logger.colors.magenta(data[item]))
                }
            }
        })
    }

    if (options.remove) {
        logger.debug('Removing alias ' + options.remove)
        instance.remove()
    }
}

Alias.prototype.add = function() {
    var instance = this,
        options = instance.options

    configs.writeGlobalConfig('alias.' + options.add, options.user)
}

Alias.prototype.list = function(opt_callback) {
    opt_callback && opt_callback(null, config.alias)
}

Alias.prototype.remove = function() {
    var instance = this,
        options = instance.options

    delete config.alias[options.remove]

    configs.writeGlobalConfig('alias', config.alias)
}

exports.Impl = Alias
