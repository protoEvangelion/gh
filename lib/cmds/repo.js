/*
 * Copyright 2013-2018, All Rights Reserved.
 *
 * Code licensed under the BSD License:
 * https://github.com/node-gh/gh/blob/master/LICENSE.md
 *
 * @author Henrique Vicente <henriquevicente@gmail.com>
 * @author Eduardo Lundgren <eduardo.lundgren@gmail.com>
 * @author Zeno Rocha <zno.rocha@gmail.com>
 */

'use strict'

// -- Requires -------------------------------------------------------------------------------------

var base = require('../base'),
    fs = require('fs'),
    git = require('../git'),
    hooks = require('../hooks'),
    logger = require('../logger'),
    openUrl = require('opn'),
    inquirer = require('inquirer'),
    url = require('url'),
    config = base.getConfig()

// -- Constructor ----------------------------------------------------------------------------------

function Repo(options) {
    this.options = options
}

// -- Constants ------------------------------------------------------------------------------------

Repo.DETAILS = {
    alias: 're',
    description: 'Provides a set of util commands to work with Repositories.',
    commands: ['browser', 'clone', 'delete', 'fork', 'list', 'new', 'update'],
    options: {
        browser: Boolean,
        clone: Boolean,
        color: String,
        delete: String,
        description: String,
        detailed: Boolean,
        gitignore: String,
        fork: String,
        homepage: String,
        init: Boolean,
        label: Boolean,
        list: Boolean,
        new: String,
        organization: String,
        page: String,
        per_page: String,
        private: Boolean,
        protocol: String,
        repo: String,
        type: ['all', 'forks', 'member', 'owner', 'public', 'private', 'sources'],
        update: String,
        user: String,
    },
    shorthands: {
        B: ['--browser'],
        c: ['--clone'],
        C: ['--color'],
        D: ['--delete'],
        d: ['--detailed'],
        f: ['--fork'],
        L: ['--label'],
        l: ['--list'],
        N: ['--new'],
        O: ['--organization'],
        p: ['--private'],
        P: ['--protocol'],
        r: ['--repo'],
        t: ['--type'],
        U: ['--update'],
        u: ['--user'],
    },
    payload: function(payload, options) {
        if (options.browser !== false) {
            options.browser = true
        }
    },
}

Repo.TYPE_ALL = 'all'
Repo.TYPE_FORKS = 'forks'
Repo.TYPE_MEMBER = 'member'
Repo.TYPE_OWNER = 'owner'
Repo.TYPE_PRIVATE = 'private'
Repo.TYPE_PUBLIC = 'public'
Repo.TYPE_SOURCES = 'sources'

// -- Commands -------------------------------------------------------------------------------------

Repo.prototype.run = function() {
    var instance = this,
        options = instance.options,
        user = options.loggedUser

    instance.config = config

    if (options.browser) {
        instance.browser(options.user, options.repo)
    }

    if (options.clone) {
        hooks.invoke('repo.get', instance, function(afterHooksCallback) {
            if (options.organization) {
                user = options.organization
            } else if (options.user) {
                user = options.user
            }

            if (fs.existsSync(process.cwd() + '/' + options.repo)) {
                logger.error(
                    "Can't clone " +
                        logger.colors.green(user + '/' + options.repo) +
                        '. ' +
                        logger.colors.green(options.repo) +
                        ' already exists in this directory.'
                )
                return
            }

            instance.get(user, options.repo, function(err1, repo) {
                if (err1) {
                    logger.error(
                        "Can't clone " +
                            logger.colors.green(user + '/' + options.repo) +
                            '. ' +
                            JSON.parse(err1).message
                    )
                    return
                }

                logger.log(repo.html_url)

                var repoUrl
                if (options.protocol) {
                    if (options.protocol === 'https') {
                        repoUrl = 'https://github.com/' + user + '/' + options.repo + '.git'
                    }
                } else {
                    repoUrl = 'git@github.com:' + user + '/' + options.repo + '.git'
                }

                if (repo) {
                    instance.clone_(user, options.repo, repoUrl)
                }

                afterHooksCallback()
            })
        })
    }

    if (options.delete && !options.label) {
        hooks.invoke('repo.delete', instance, function(afterHooksCallback) {
            logger.log('Deleting repo ' + logger.colors.green(options.user + '/' + options.delete))

            inquirer
                .prompt([
                    {
                        type: 'input',
                        message: 'Are you sure? This action CANNOT be undone. [y/N]',
                        name: 'confirmation',
                    },
                ])
                .then(function(answers) {
                    if (answers.confirmation.toLowerCase() === 'y') {
                        instance.delete(options.user, options.delete, function(err) {
                            if (err) {
                                logger.error("Can't delete repo.")
                                return
                            }

                            afterHooksCallback()
                        })
                    } else {
                        logger.log('Not deleted.')
                    }
                })
        })
    }

    if (options.fork) {
        hooks.invoke('repo.fork', instance, function(afterHooksCallback) {
            if (options.organization) {
                user = options.organization
            }

            options.repo = options.fork

            logger.log(
                'Forking repo ' +
                    logger.colors.green(options.user + '/' + options.repo) +
                    ' on ' +
                    logger.colors.green(user + '/' + options.repo)
            )

            instance.fork(function(err1, repo) {
                if (err1) {
                    logger.error("Can't fork. " + JSON.parse(err1).message)
                    return
                }

                logger.log(repo.html_url)

                if (repo && options.clone) {
                    instance.clone_(options.loggedUser, options.repo, repo.ssh_url)
                }

                afterHooksCallback()
            })
        })
    }

    if (options.label) {
        if (options.organization) {
            user = options.organization
        } else if (options.user) {
            user = options.user
        }

        if (options.delete) {
            hooks.invoke('repo.deleteLabel', instance, function(afterHooksCallback) {
                options.label = options.delete

                logger.log(
                    'Deleting label ' +
                        logger.colors.green(options.label) +
                        ' on ' +
                        logger.colors.green(user + '/' + options.repo)
                )

                instance.deleteLabel(user, function(err1) {
                    if (err1) {
                        logger.error("Can't delete label.")
                        return
                    }

                    afterHooksCallback()
                })
            })
        } else if (options.list) {
            hooks.invoke('repo.listLabels', instance, function(afterHooksCallback) {
                if (options.page) {
                    logger.log(
                        'Listing labels from page ' +
                            logger.colors.green(options.page) +
                            ' on ' +
                            logger.colors.green(user + '/' + options.repo)
                    )
                } else {
                    logger.log(
                        'Listing labels' + ' on ' + logger.colors.green(user + '/' + options.repo)
                    )
                }

                instance.listLabels(user, function(err1) {
                    if (err1) {
                        logger.error("Can't list labels.")
                        return
                    }

                    afterHooksCallback()
                })
            })
        } else if (options.new) {
            hooks.invoke('repo.createLabel', instance, function(afterHooksCallback) {
                options.label = options.new

                logger.log(
                    'Creating label ' +
                        logger.colors.green(options.label) +
                        ' on ' +
                        logger.colors.green(user + '/' + options.repo)
                )

                instance.createLabel(user, function(err1) {
                    if (err1) {
                        logger.error("Can't create label.")
                        return
                    }

                    afterHooksCallback()
                })
            })
        } else if (options.update) {
            hooks.invoke('repo.updateLabel', instance, function(afterHooksCallback) {
                options.label = options.update

                logger.log(
                    'Updating label ' +
                        logger.colors.green(options.label) +
                        ' on ' +
                        logger.colors.green(user + '/' + options.repo)
                )

                instance.updateLabel(user, function(err1) {
                    if (err1) {
                        logger.error("Can't update label.")
                        return
                    }

                    afterHooksCallback()
                })
            })
        }
    }

    if (options.list && !options.label) {
        if (options.organization) {
            user = options.organization
            options.type = options.type || Repo.TYPE_ALL
        } else {
            user = options.user
            options.type = options.type || Repo.TYPE_OWNER
        }

        if (options.isTTY.out) {
            logger.log(
                'Listing ' +
                    logger.colors.green(options.type) +
                    ' repos for ' +
                    logger.colors.green(user)
            )
        }

        instance.list(user, function(err) {
            if (err) {
                logger.error("Can't list repos.")
            }
        })
    }

    if (options.new && !options.label) {
        hooks.invoke('repo.new', instance, function(afterHooksCallback) {
            options.repo = options.new

            if (options.organization) {
                options.user = options.organization
            }

            logger.log(
                'Creating a new repo on ' + logger.colors.green(options.user + '/' + options.new)
            )

            instance.new(function(err1, repo) {
                if (err1) {
                    logger.error("Can't create new repo. " + JSON.parse(err1.message).message)
                    return
                }

                logger.log(repo.html_url)

                if (repo && options.clone) {
                    instance.clone_(options.user, options.repo, repo.ssh_url)
                }

                afterHooksCallback()
            })
        })
    }
}

Repo.prototype.browser = function(user, repo) {
    openUrl(config.github_host + user + '/' + repo, { wait: false })
}

Repo.prototype.clone_ = function(user, repo, repo_url) {
    logger.log('Cloning ' + logger.colors.green(user + '/' + repo))
    git.clone(url.parse(repo_url).href, repo)
}

Repo.prototype.createLabel = function(user, opt_callback) {
    var instance = this,
        options = instance.options,
        payload

    payload = {
        color: options.color,
        name: options.new,
        repo: options.repo,
        user: user,
    }

    base.github.issues.createLabel(payload, opt_callback)
}

Repo.prototype.delete = function(user, repo, opt_callback) {
    var payload

    payload = {
        user: user,
        repo: repo,
    }

    base.github.repos.delete(payload, opt_callback)
}

Repo.prototype.deleteLabel = function(user, opt_callback) {
    var instance = this,
        options = instance.options,
        payload

    payload = {
        name: options.delete,
        repo: options.repo,
        user: user,
    }

    base.github.issues.deleteLabel(payload, opt_callback)
}

Repo.prototype.get = function(user, repo, opt_callback) {
    var payload

    payload = {
        user: user,
        repo: repo,
    }

    base.github.repos.get(payload, opt_callback)
}

Repo.prototype.list = function(user, opt_callback) {
    var instance = this,
        method = 'getFromUser',
        options = instance.options,
        payload

    payload = {
        type: options.type,
        user: user,
        per_page: 100,
    }

    if (options.organization) {
        method = 'getFromOrg'
        payload.org = options.organization
    }

    if (options.type === 'public' || options.type === 'private') {
        if (user === options.user) {
            method = 'getAll'
        } else {
            logger.error('You can only list your own public and private repos.')
            return
        }
    }

    base.github.repos[method](payload, function(err, repos) {
        instance.listCallback_(err, repos, opt_callback)
    })
}

Repo.prototype.listCallback_ = function(err, repos, opt_callback) {
    var instance = this,
        options = instance.options,
        pos,
        repo

    if (err && !options.all) {
        logger.error(logger.getErrorMessage(err))
    }

    if (repos && repos.length > 0) {
        for (pos in repos) {
            if (repos.hasOwnProperty(pos) && repos[pos].full_name) {
                repo = repos[pos]
                logger.log(repo.full_name)

                if (options.detailed) {
                    logger.log(logger.colors.blue(repo.html_url))

                    if (repo.description) {
                        logger.log(logger.colors.blue(repo.description))
                    }

                    if (repo.homepage) {
                        logger.log(logger.colors.blue(repo.homepage))
                    }

                    logger.log('last update ' + logger.getDuration(repo.updated_at))
                }

                if (options.isTTY.out) {
                    logger.log(
                        logger.colors.green(
                            'forks: ' +
                                repo.forks +
                                ', stars: ' +
                                repo.watchers +
                                ', issues: ' +
                                repo.open_issues
                        ) + '\n'
                    )
                }
            }
        }

        opt_callback && opt_callback(err)
    }
}

Repo.prototype.listLabels = function(user, opt_callback) {
    var instance = this,
        options = instance.options,
        payload

    payload = {
        page: options.page,
        per_page: options.per_page,
        repo: options.repo,
        user: user,
    }

    base.github.issues.getLabels(payload, function(err, labels) {
        instance.listLabelsCallback_(err, labels, opt_callback)
    })
}

Repo.prototype.listLabelsCallback_ = function(err, labels, opt_callback) {
    var instance = this,
        options = instance.options

    if (err && !options.all) {
        logger.error(logger.getErrorMessage(err))
    }

    if (labels && labels.length > 0) {
        labels.forEach(function(label) {
            logger.log(logger.colors.yellow(label.name))
        })

        opt_callback && opt_callback(err)
    }
}

Repo.prototype.fork = function(opt_callback) {
    var instance = this,
        options = instance.options,
        payload

    payload = {
        user: options.user,
        repo: options.repo,
    }

    if (options.organization) {
        payload.organization = options.organization
    }

    base.github.repos.fork(payload, opt_callback)
}

Repo.prototype.new = function(opt_callback) {
    var instance = this,
        options = instance.options,
        payload,
        method = 'create'

    options.description = options.description || ''
    options.gitignore = options.gitignore || ''
    options.homepage = options.homepage || ''
    options.init = options.init || false

    if (options.type === Repo.TYPE_PRIVATE) {
        options.private = true
    }

    options.private = options.private || false

    if (options.gitignore) {
        options.init = true
    }

    payload = {
        auto_init: options.init,
        description: options.description,
        gitignore_template: options.gitignore,
        homepage: options.homepage,
        name: options.new,
        private: options.private,
    }

    if (options.organization) {
        method = 'createFromOrg'
        payload.org = options.organization
    }

    base.github.repos[method](payload, opt_callback)
}

Repo.prototype.updateLabel = function(user, opt_callback) {
    var instance = this,
        options = instance.options,
        payload

    payload = {
        color: options.color,
        name: options.update,
        repo: options.repo,
        user: user,
    }

    base.github.issues.updateLabel(payload, opt_callback)
}

exports.Impl = Repo
