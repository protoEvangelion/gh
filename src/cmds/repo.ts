/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

// -- Requires -------------------------------------------------------------------------------------

import * as Octokit from '@octokit/rest'
import * as fs from 'fs'
import * as inquirer from 'inquirer'
import { openUrl, getCurrentFolderName, userRanValidFlags } from '../utils'
import * as url from 'url'
import * as base from '../base'
import * as git from '../git'

import { afterHooks, beforeHooks } from '../hooks'
import * as logger from '../logger'

const config = base.getConfig()

// -- Constants ------------------------------------------------------------------------------------

export const name = 'Repo'
export const DETAILS = {
    alias: 're',
    description: 'Provides a set of util commands to work with Repositories.',
    commands: ['browser', 'clone', 'delete', 'fork', 'list', 'new', 'update'],
    options: {
        browser: Boolean,
        clone: Boolean,
        color: String,
        date: String,
        delete: String,
        description: String,
        detailed: Boolean,
        gitignore: String,
        fork: String,
        homepage: String,
        init: Boolean,
        label: Boolean,
        list: Boolean,
        new: String,
        organization: String,
        page: String,
        per_page: String,
        private: Boolean,
        protocol: String,
        repo: String,
        type: ['all', 'forks', 'member', 'owner', 'public', 'private', 'source'],
        update: String,
        user: String,
    },
    shorthands: {
        B: ['--browser'],
        c: ['--clone'],
        C: ['--color'],
        D: ['--delete'],
        d: ['--detailed'],
        f: ['--fork'],
        L: ['--label'],
        l: ['--list'],
        N: ['--new'],
        O: ['--organization'],
        p: ['--private'],
        P: ['--protocol'],
        r: ['--repo'],
        t: ['--type'],
        U: ['--update'],
        u: ['--user'],
    },
}

const TYPE_ALL = 'all'
const TYPE_FORKS = 'forks'
const TYPE_MEMBER = 'member'
const TYPE_OWNER = 'owner'
const TYPE_PRIVATE = 'private'
const TYPE_PUBLIC = 'public'
const TYPE_SOURCES = 'sources'

// -- Commands -------------------------------------------------------------------------------------

export async function run(options, done) {
    let user = options.loggedUser

    config = config

    if (
        !userRanValidFlags(DETAILS.commands, options) &&
        options.browser !== false &&
        options.argv.cooked.length === 1
    ) {
        options.browser = true
    }

    if (options.browser) {
        browser(options.user, options.repo)
    } else if (options.clone && !options.new) {
        await beforeHooks('repo.get', { options })

        if (options.organization) {
            user = options.organization
        } else if (options.user) {
            user = options.user
        }

        if (fs.existsSync(`${process.cwd()}/${options.repo}`)) {
            logger.error(
                `Can't clone ${logger.colors.green(
                    `${user}/${options.repo}`
                )}. ${logger.colors.green(options.repo)} already exists in this directory.`
            )
            return
        }

        try {
            var { data } = await get(user, options.repo)
        } catch (err) {
            throw new Error(
                `Can't clone ${logger.colors.green(`${user}/${options.repo}`)}. ${
                    JSON.parse(err).message
                }\n${err}`
            )
        }

        logger.log(data.html_url)

        let repoUrl

        if (options.protocol) {
            if (options.protocol === 'https') {
                repoUrl = `https://github.com/${user}/${options.repo}.git`
            }
        } else {
            repoUrl = `git@github.com:${user}/${options.repo}.git`
        }

        if (data) {
            clone_(user, options.repo, repoUrl)
        }

        await afterHooks('repo.get', { options })
    } else if (options.delete && !options.label) {
        await beforeHooks('repo.delete', { options })

        logger.log(`Deleting repo ${logger.colors.green(`${options.user}/${options.delete}`)}`)

        const answers = await inquirer.prompt([
            {
                type: 'input',
                message: 'Are you sure? This action CANNOT be undone. [y/N]',
                name: 'confirmation',
            },
        ])

        if (answers.confirmation.toLowerCase() === 'y') {
            try {
                const { status } = await deleteRepo(options.user, options.delete)

                status === 204 && logger.log('Successfully deleted repo.')
            } catch (err) {
                logger.error(`Can't delete repo.\n${err}`)
            }

            await afterHooks('repo.delete', { options })
        } else {
            logger.log('Not deleted.')
        }
    } else if (options.fork) {
        await beforeHooks('repo.fork', { options })

        if (options.organization) {
            user = options.organization
        }

        options.repo = options.fork

        logger.log(
            `Forking repo ${logger.colors.green(
                `${options.user}/${options.repo}`
            )} into ${logger.colors.green(`${user}/${options.repo}`)}`
        )

        try {
            var { data } = await fork()
        } catch (err) {
            throw new Error(`Can't fork. ${err}`)
        }

        logger.log(`Successfully forked: ${data.html_url}`)

        if (data && options.clone) {
            clone_(user, options.repo, data.ssh_url)
        }

        await afterHooks('repo.fork', { options })
    } else if (options.label) {
        if (options.organization) {
            user = options.organization
        } else if (options.user) {
            user = options.user
        }

        if (options.delete) {
            await beforeHooks('repo.deleteLabel', { options })

            options.label = options.delete

            logger.log(
                `Deleting label ${logger.colors.green(options.label)} on ${logger.colors.green(
                    `${user}/${options.repo}`
                )}`
            )

            try {
                var { status } = await deleteLabel(user)
            } catch (err) {
                throw new Error(`Can't delete label.\n${err}`)
            }

            status === 204 && logger.log('Successful.')

            await afterHooks('repo.deleteLabel', { options })
        } else if (options.list) {
            await beforeHooks('repo.listLabels', { options })

            if (options.page) {
                logger.log(
                    `Listing labels from page ${logger.colors.green(
                        options.page
                    )} on ${logger.colors.green(`${user}/${options.repo}`)}`
                )
            } else {
                logger.log(`Listing labels on ${logger.colors.green(`${user}/${options.repo}`)}`)
            }

            try {
                var { data } = await listLabels(user)
            } catch (err) {
                throw new Error(`Can't list labels\n${err}`)
            }

            data.forEach(label => logger.log(logger.colors.yellow(label.name)))

            await afterHooks('repo.listLabels', { options })
        } else if (options.new) {
            await beforeHooks('repo.createLabel', { options })

            options.label = options.new

            logger.log(
                `Creating label ${logger.colors.green(options.label)} on ${logger.colors.green(
                    `${user}/${options.repo}`
                )}`
            )

            try {
                await createLabel(user)
            } catch (err) {
                throw new Error(`Can't create label.\n${err}`)
            }

            await afterHooks('repo.createLabel', { options })
        } else if (options.update) {
            await beforeHooks('repo.updateLabel', { options })

            options.label = options.update

            logger.log(
                `Updating label ${logger.colors.green(options.label)} on ${logger.colors.green(
                    `${user}/${options.repo}`
                )}`
            )

            try {
                var { status } = await updateLabel(user)
            } catch (err) {
                throw new Error(`Can't update label.\n${err}`)
            }

            status === 200 && logger.log('Success')

            await afterHooks('repo.updateLabel', { options })
        }
    } else if (options.list && !options.label) {
        await beforeHooks('repo.list', { options })

        if (options.organization) {
            user = options.organization
            options.type = options.type || Repo.TYPE_ALL
        } else {
            user = options.user
            options.type = options.type || Repo.TYPE_OWNER
        }

        // Add a isTTY value on the options to determine whether or not the command is executed in a TTY context or not.
        // Will be false if cmd is piped like: gh re --list | cat
        if (Boolean(process.stdout.isTTY)) {
            logger.log(
                `Listing ${logger.colors.green(options.type)} repos for ${logger.colors.green(
                    user
                )}`
            )
        }
        try {
            var data = await list(user)
        } catch (err) {
            throw new Error(`Can't list repos.\n${err}`)
        }

        listCallback_(data)

        await afterHooks('repo.list', { options })
    } else if (options.new && !options.label) {
        if (!options.new.trim()) {
            options.new = getCurrentFolderName()
        }

        await beforeHooks('repo.new', { options })

        options.repo = options.new

        if (options.organization) {
            options.user = options.organization
        }

        logger.log(
            `Creating a new repo on ${logger.colors.green(`${options.user}/${options.new}`)}`
        )

        try {
            var { data } = await newRepo(options)
        } catch (err) {
            throw new Error(`Can't create new repo.\n${err}`)
        }

        logger.log(data.html_url)

        if (data && options.clone) {
            clone_(options.user, options.repo, data.ssh_url)
        }

        await afterHooks('repo.new', { options })
    }

    done && done()
}

function browser(user, repo) {
    openUrl(`${config.github_host}/${user}/${repo}`)
}

function clone_(user, repo, repo_url) {
    logger.log(`Cloning ${logger.colors.green(`${user}/${repo}`)}`)
    git.clone(url.parse(repo_url).href, repo)
}

function createLabel(options, user): Promise<Octokit.IssuesCreateLabelResponse> {
    const payload: Octokit.IssuesCreateLabelParams = {
        owner: user,
        color: normalizeColor(options.color),
        name: options.new,
        repo: options.repo,
    }

    if (options.description) {
        payload.description = options.description
    }

    return options.GitHub.issues.createLabel(payload)
}

function deleteRepo(user, repo): Promise<Octokit.ReposDeleteResponse> {
    const payload = {
        repo,
        owner: user,
    }

    return options.GitHub.repos.delete(payload)
}

function deleteLabel(options, user): Promise<Octokit.IssuesDeleteLabelResponse> {
    const payload = {
        owner: user,
        name: options.delete,
        repo: options.repo,
    }

    return options.GitHub.issues.deleteLabel(payload)
}

function get(user, repo): Promise<Octokit.IssuesGetResponse> {
    const payload = {
        repo,
        owner: user,
    }

    return options.GitHub.repos.get(payload)
}

function list(options, user): Promise<Octokit.AnyResponse | Octokit.ReposListForOrgResponse> {
    let method = 'listForUser'

    const payload: any = {
        type: options.type,
        per_page: 100,
    }

    if (options.organization) {
        method = 'listForOrg'
        payload.org = options.organization
    } else {
        payload.username = options.user
    }

    if (options.type === 'public' || options.type === 'private') {
        if (user === options.user) {
            method = 'listForUser'
        } else {
            logger.error('You can only list your own public and private repos.')
            return
        }
    }

    return options.GitHub.paginate(options.GitHub.repos[method].endpoint(payload))
}

function listCallback_(options, repos): void {
    let pos
    let repo

    if (repos && repos.length > 0) {
        for (pos in repos) {
            if (repos.hasOwnProperty(pos) && repos[pos].full_name) {
                repo = repos[pos]
                logger.log(repo.full_name)

                if (options.detailed) {
                    logger.log(logger.colors.blue(repo.html_url))

                    if (repo.description) {
                        logger.log(logger.colors.blue(repo.description))
                    }

                    if (repo.homepage) {
                        logger.log(logger.colors.blue(repo.homepage))
                    }

                    logger.log(`last update ${logger.getDuration(repo.updated_at, options.date)}`)
                }

                if (Boolean(process.stdout.isTTY)) {
                    logger.log(
                        `${logger.colors.green(
                            `forks: ${repo.forks}, stars: ${repo.watchers}, issues: ${
                                repo.open_issues
                            }`
                        )}\n`
                    )
                }
            }
        }
    }
}

function listLabels(options, user): Promise<Octokit.IssuesListLabelsForRepoResponse> {
    const payload: Octokit.IssuesListLabelsForRepoParams = {
        owner: user,
        repo: options.repo,
        ...(options.page && { page: options.page }),
        ...(options.per_page && { per_page: options.per_page }),
    }

    return options.GitHub.issues.listLabelsForRepo(payload)
}

function listLabelsCallback_(options, err, labels): void {
    if (err && !options.all) {
        logger.error(logger.getErrorMessage(err))
    }

    if (labels && labels.length > 0) {
        labels.forEach(label => {
            logger.log(logger.colors.yellow(label.name))
        })
    }
}

async function fork(options): Promise<Octokit.ReposCreateForkResponse> {
    const payload: Octokit.ReposCreateForkParams = {
        owner: options.user,
        repo: options.repo,
    }

    if (options.organization) {
        payload.organization = options.organization
    }

    return await options.GitHub.repos.createFork(payload)
}

function newRepo(
    options
): Promise<Octokit.ReposCreateInOrgResponse | Octokit.ReposCreateForAuthenticatedUserResponse> {
    let method = 'createForAuthenticatedUser'

    options.description = options.description || ''
    options.gitignore = options.gitignore || ''
    options.homepage = options.homepage || ''
    options.init = options.init || false

    if (options.type === Repo.TYPE_PRIVATE) {
        options.private = true
    }

    options.private = options.private || false

    if (options.gitignore) {
        options.init = true
    }

    const payload: any = {
        auto_init: options.init,
        description: options.description,
        gitignore_template: options.gitignore,
        homepage: options.homepage,
        name: options.new,
        private: options.private,
    }

    if (options.organization) {
        method = 'createInOrg'
        payload.org = options.organization
    }

    return options.GitHub.repos[method](payload)
}

function updateLabel(options, user): Promise<Octokit.IssuesUpdateLabelResponse> {
    const payload: Octokit.IssuesUpdateLabelParams = {
        owner: user,
        color: normalizeColor(options.color),
        current_name: options.update,
        repo: options.repo,
    }

    return options.GitHub.issues.updateLabel(payload)
}

function normalizeColor(color) {
    return color.includes('#') ? color.replace('#', '') : color
}
