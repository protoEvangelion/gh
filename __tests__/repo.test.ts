/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as stripAnsi from 'strip-ansi'
import { runCmd } from '../src/test-utils'
import { getCloneUrl } from '../src/cmds/repo'

describe('E2E: Repo Module Test', () => {
    it('List Repos `gh re --list`', done => {
        expect(runCmd('gh re --list')).toMatchSnapshot()
        done()
    })

    it('List Repos `gh re --browser`', done => {
        expect(runCmd('gh re --browser')).toMatchSnapshot()
        done()
    })

    it('Search repositories from GitHub org showing detailed results. `gh re --detailed --search octocat org:github`', done => {
        expect(runCmd('gh re --detailed --search octocat org:github')).toMatchSnapshot()
        done()
    })

    it('Create new repo `gh re --new foo --init`', done => {
        expect(runCmd('gh re --new foo --init')).toMatchSnapshot()
        done()
    })

    it('List labels on a repo `gh re --label --list --repo gh`', done => {
        expect(runCmd('gh re --label --list --repo gh')).toMatchSnapshot()
        done()
    })

    it('Fork a repo `gh re --fork prettier --user prettier`', done => {
        expect(runCmd('gh re --fork prettier --user prettier')).toMatchSnapshot()
        done()
    })

    it('Delete repo `gh re --delete foo`', done => {
        expect(stripAnsi(runCmd('printf "y" | gh re --delete foo'))).toMatchSnapshot()
        done()
    })
})

describe('Unit tests for repo cmd', () => {
    it('Cloning repo url is correct', () => {
        const options = { repo: 'gh', user: 'node-gh', github_host: 'https://github.com' }
        expect(getCloneUrl(options)).toEqual('git@github.com:node-gh/gh.git')
        expect(getCloneUrl({ ...options, protocol: 'https' })).toEqual(
            'https://github.com/node-gh/gh.git'
        )
        expect(getCloneUrl({ ...options, protocol: 'https' }, 'custom-ssh-host')).toEqual(
            'https://github.com/node-gh/gh.git'
        )
        expect(getCloneUrl(options, 'custom-ssh-host')).toEqual(
            'git@custom-ssh-host:node-gh/gh.git'
        )
    })
})
