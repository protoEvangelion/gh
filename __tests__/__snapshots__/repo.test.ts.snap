// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E: Repo Module Test Create new repo \`gh re --new foo --init\` 1`] = `
"Creating a new repo on protoEvangelion/foo
https://github.com/protoEvangelion/foo
[hook] gh re --browser  --user protoEvangelion --repo foo
"
`;

exports[`E2E: Repo Module Test Delete repo \`gh re --delete foo\` 1`] = `
"Deleting repo protoEvangelion/foo
? Are you sure? This action CANNOT be undone. [y/N] ? Are you sure? This action CANNOT be undone. [y/N] y? Are you sure? This action CANNOT be undone. [y/N] y
âœ“ Successfully deleted repo.
"
`;

exports[`E2E: Repo Module Test Fork a repo \`gh re --fork prettier --user prettier\` 1`] = `
"Forking repo prettier/prettier into protoEvangelion/prettier
Successfully forked: https://github.com/protoEvangelion/prettier
"
`;

exports[`E2E: Repo Module Test List Repos \`gh re --browser\` 1`] = `
"https://github.com/protoEvangelion/gh
"
`;

exports[`E2E: Repo Module Test List Repos \`gh re --list\` 1`] = `
"protoEvangelion/100-days-of-code
protoEvangelion/adopt-a-family
protoEvangelion/alloy-ui
"
`;

exports[`E2E: Repo Module Test List labels on a repo \`gh re --label --list --repo gh\` 1`] = `
"Listing labels on protoEvangelion/gh
--user
duplicate
enhancement
"
`;

exports[`E2E: Repo Module Test Search repositories from GitHub org showing detailed results. \`gh re --detailed --search octocat org:github\` 1`] = `
"Searching for repos using the criteria: octocat org:github
github/octocatalog-diff
https://github.com/github/octocatalog-diff
Compile Puppet catalogs from 2 branches, versions, etc., and compare them
https://rubygems.org/gems/octocatalog-diff/
last update in 4 years
"
`;
